src/compute/src/sink/persist_sink.rs

```diff
@@ -1181,3 +1190,86 @@ where
     let token = Rc::new(shutdown_button.press_on_drop());
     (output_stream, token)
 }
+
+/// This is for REFRESH options on materialized views. It adds an operator that rounds up the
```

Comment:
My gut feeling is that this operator is too complex and does multiple things at once, which we might want to separate out. It rounds up the timestamp (this could be a simple `map`), it delays the data (there's the `delay` operator), and it eventually ceases producing data. I think strongly that it should be factored in this way to make it clear what the operator is doing, and to avoid the pitfalls it currently has.

I think the property that we'd want to close this sink at some point needs to be encoded in the `up_to` of the dataflow itself, and not here.

Reply:
Correct me if I'm wrong, but I don't think the `delay` operator can do what we need to do here. Note that we don't just need to delay the data, but we also have to make sure that the operators frontier advances to the next refresh interval as soon as it has produced all outputs for the current interval. This is important because otherwise the MV wouldn't remain queryable (under strict serializable) while we take the replica down between refresh intervals.

AIUI, to make the operators frontier jump to the future, you need to fiddle with the input path summaries somehow. The `delay` operator doesn't do that, so it's output frontier will always be bound by the capabilities still held by upstream operators, which would prevent it from jumping into the future.

Reply:
We'll discuss this next week.

Btw. this is the relevant part of the design doc: https://github.com/ggevay/materialize/blob/refresh-every-design-doc/doc/developer/design/20231027_refresh_every_mv.md#b-rounding-up-timestamps-going-into-the-persist-sink

Reply:
Yup, the delay operator doesn't have any special connectivity, so we'd need to roll our own.

Reply:
I've refactored the operator as discussed in Timely office hours.